cpp/src/arrow/util/future.h

```diff
@@ -829,6 +829,9 @@ Future<std::vector<Result<T>>> All(std::vector<Future<T>> futures) {
   return out;
 }
 
+template <>
+inline Future<>::Future(Status s) : Future(internal::Empty::ToResult(std::move(s))) {}
```

Comment:
This is because the usual implicit constructor for `Future<Status>` delegates to `Future(Result<ValueType>(Status))` which errors on an OK Status (the assumption being that an OK Status will never arise); however for `Future<>` we actually do want to handle an OK Status.

It's not inline because it's a specialization for `Future<>`. I tried to use `enable_if` to do it inline but couldn't figure out the right template magic needed to get it to work.

Reply:
Ah, thank you.